relay_rpc/src/rpc/params/session.rs

```diff
@@ -0,0 +1,569 @@
+
+impl Deref for ProposeNamespaces {
+    type Target = BTreeMap<String, ProposeNamespace>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl ProposeNamespaces {
+    /// Ensures that application is compatible with the requester requirements.
+    ///
+    /// Implementation must support at least all the elements in `required`.
+    pub fn supported(&self, required: &ProposeNamespaces) -> Result<(), ProposeNamespaceError> {
+        if self.is_empty() {
+            return Err(ProposeNamespaceError::UnsupportedNamespace(
+                "None supported".to_string(),
+            ));
+        }
+
+        if required.is_empty() {
+            return Ok(());
+        }
+
+        for (name, other) in required.iter() {
+            let ours = self
+                .get(name)
+                .ok_or_else(|| ProposeNamespaceError::UnsupportedNamespace(name.to_string()))?;
+            ours.supported(other)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn caip2_validate(&self) -> Result<(), ProposeNamespaceError> {
+        let caip_regex = get_caip2_regex();
+        for (name, namespace) in self.deref() {
+            let captures = caip_regex
+                .captures(name)
+                .ok_or_else(|| ProposeNamespaceError::UnsupportedNamespaceKey(name.to_string()))?;
+
+            let name = captures
+                .name("namespace")
+                .expect("namespace name missing: unexpected error")
+                .as_str();
+
+            let reference = captures.name("reference").map(|m| m.as_str());
+
+            namespace.chains_caip2_validate(name, reference)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// TODO: some validation from `ProposeNamespaces` should be re-used.
+/// TODO: caip-10 validation.
+/// TODO: named errors like in `ProposeNamespaces`.
+#[derive(Debug, Serialize, PartialEq, Eq, Hash, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct SettleNamespaces(pub BTreeMap<String, Namespace>);
+
+impl Deref for SettleNamespaces {
+    type Target = BTreeMap<String, Namespace>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Serialize, PartialEq, Eq, Hash, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct SettleNamespace {
+    pub accounts: BTreeSet<String>,
+    pub methods: BTreeSet<String>,
+    pub events: BTreeSet<String>,
+}
+
+#[derive(Debug, Serialize, PartialEq, Eq, Hash, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Namespace {
+    pub chains: Option<BTreeSet<String>>,
+    pub accounts: Option<BTreeSet<String>>,
+    pub methods: BTreeSet<String>,
+    pub events: BTreeSet<String>,
+}
+
+// Trims json of the whitespaces and newlines.
+/// Allows to use "pretty json" in unittest, and still get consistent
+/// results post serialization/deserialization.
+#[cfg(test)]
+pub fn param_json_trim(json: &str) -> String {
```

Comment:
I think 'trim' usually means remove chars from beginning or end.
Also, this fn does not do anything specific to json.
Why not just `remove_whitespaces()`?

Reply:
we also need to handle whitespaces/linebreaks in between chars too not just beginning/end. So I guess remove_whitespaces() won't be sufficient.